### PR TITLE
Prototypical batching for GA

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Related to #1402 (but not fully addressed yet)

##### What are the significant changes in functionality due to this change request?
This PR introduce an initial prototypical method to execute jobs (models' evaluations) in batches. 
This methodology allows Optimizers and Samplers to submit a set of realizations in "batches". The JobHandler
will execute all the jobs and return them (in bulk) once all of them are finished. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.  *Tests will be added with later commit.*
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by alfoa** [comment]
@joshua-cogliati-inl @mandd @wangcj05

**Comment by alfoa** [comment]
I am still running the tests locally to check if they work but this can be start to be reviewed (from the code prospective)

**Comment by alfoa** [comment]
@joshua-cogliati-inl @mandd this now works.

**Comment by mandd** [comment]
> @joshua-cogliati-inl @mandd this now works.

Awesome, thanks!

**Comment by alfoa** [comment]
@mandd @joshua-cogliati-inl  I think this PR is ready to go (or at least to be reviewed). I think I can modify the "allMathcing" thing directly in the GA branch... At least it is testable in the actualy algorithm

**Review by joshua-cogliati-inl — CHANGES_REQUESTED** [review]
I have some questions and some comments on the pull request.  Thank you for splitting it out, it is more understandable.

File: framework/DataObjects/DataSet.py [hunk 1]
```diff
@@ -1575,41 +1568,30 @@ def _getRealizationFromCollectorByIndex(self,index):
...
     """
       Obtains a realization from the collector storage matching the provided index
       @ In, toMatch, dict, elements to match
       @ In, noMatch, dict, elements to AVOID matching (should not match within tolerance)
       @ In, tol, float, optional, tolerance to which match should be made
-      @ In, options, dict, optional, options to be applied to the search
-      @ Out, r, int, index where match was found OR size of data if not found
-      @ Out, rlz, dict, realization as {var:value} OR None if not found
+      @ In, first, bool, optional, return the first matching realization only? Default:True
+      @ Out, r (or rr), int or list, if first: int, index where match was found OR size of data if not found
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/DataObjects/DataSet.py:1578]: fixed

File: framework/DataObjects/DataSet.py [hunk 2]
```diff
@@ -1575,41 +1568,30 @@ def _getRealizationFromCollectorByIndex(self,index):
...
+                             else    : list, list of indeces where matches were found OR size of data if not found
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/DataObjects/DataSet.py:1579]: fixed

File: framework/Steps.py [hunk 3]
```diff
@@ -679,82 +679,131 @@ def _localTakeAstepRun(self,inDictionary):
...
+        (i.e., a DataObject, File, or HDF5, I guess? Maybe these should all
+        inherit from some base "Data" so that we can ensure a consistent
+        interface for these?)
+      @ In, outputs, object, the raven object used as the output in this step
+        (i.e., a DataObject, File, or HDF5, I guess? Maybe these should all
+        inherit from some base "Data" so that we can ensure a consistent
+        interface for these?)
+      @ In, jobHandler, object, the raven object used to handle jobs
+      @ In, inDictionary, dict, additional step objects map
+      @ In, verbose, bool, optional, if True print DEBUD statements
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Steps.py:780]: fixed

File: framework/JobHandler.py [hunk 4]
```diff
@@ -530,39 +556,60 @@ def getFinished(self, removeFinished=True, jobIdentifier = '', uniqueHandler = "
         each runner. If provided, just the jobs that have the uniqueIdentifier
         will be retrieved. By default uniqueHandler = 'any' => all the jobs for
         which no uniqueIdentifier has been set up are going to be retrieved
-      @ Out, finished, list, list of finished jobs (InternalRunner or
+      @ Out, finished, list, list of list containing finished jobs (InternalRunner or
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/JobHandler.py:559]: commented

File: framework/JobHandler.py [hunk 5]
```diff
@@ -530,39 +556,60 @@ def getFinished(self, removeFinished=True, jobIdentifier = '', uniqueHandler = "
...
+
+        ##Since these indices are sorted, reverse them to ensure that when we
+        ## delete something it will not shift anything to the left (lower index)
+        ## than it.
+      if removeFinished:
+        for i in reversed(runsToBeRemoved):
+          self.__finished[i].trackTime('collected')
+          try:
+            del self.__finished[i]
+          except ImportError:
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/JobHandler.py:609]: I think it was a debug statement.... removed

[the remark this answers is not in the document]
Comment (alfoa) [reply framework/JobHandler.py:609]: I think it was a debug statement...removed

File: framework/JobHandler.py [hunk 6]
```diff
@@ -334,6 +340,17 @@ def addJob(self, args, functionToRun, identifier, metadata=None, forceUseThreads
 
     # set the client info
     internalJob.clientRunner = clientQueue
+    #  set the groupping id if present
+    if groupInfo is not None:
+      groupId =  groupInfo['id']
+      # TODO: create method in Runner to set flags,ids,etc in the instanciated runner
+      internalJob.groupId = groupId
+      if groupId not in self.__batching:
+        self.__batching[groupId] = {"counter": 0, "ids": [], "size": groupInfo['size'], 'finished': []}
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/JobHandler.py:349]: documented

File: framework/Steps.py [hunk 7]
```diff
@@ -679,82 +679,131 @@ def _localTakeAstepRun(self,inDictionary):
...
-        self.counter +=1
-        # collect run if it succeeded
-        if finishedJob.getReturnCode() == 0:
-          for myLambda, outIndex in self._outputCollectionLambda:
-            myLambda([finishedJob,outputs[outIndex]])
-            self.raiseADebug('Just collected job {j:^8} and sent to output "{o}"'
+
+      ##BATCH... TO MODIFY. FIXME
+      for finishedJobObjs in finishedJobs:
+        if type(finishedJobObjs).__name__ in 'list':
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Steps.py:685]: commented

File: framework/Steps.py [hunk 8]
```diff
@@ -679,82 +679,131 @@ def _localTakeAstepRun(self,inDictionary):
...
+                  self.failedRuns.append(copy.copy(finishedJob))
+                self.raiseAWarning('The job "'+finishedJob.identifier+'" has been submitted '+ str(self.failureHandling['repetitions'])+' times, failing all the times!!!')
+            if sampler.ableToHandelFailedRuns:
+              self.raiseAWarning('The sampler/optimizer "'+sampler.type+'" is able to handle failed runs!')
+            #pop the failed job from the list
+            finishedJobList.pop(finishedJobList.index(finishedJob))
+        if type(finishedJobObjs).__name__ in 'list': # TODO: should be consistent, if no batching should batch size be 1 or 0 ?
+          # if sampler claims it's batching, then only collect once, since it will collect the batch
+          # together, not one-at-a-time
+          sampler.finalizeActualSampling(finishedJobs[0][0],model,inputs)
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Steps.py:730]: commented

**Comment by alfoa** [comment]
@joshua-cogliati-inl  I pushed the modifications

File: framework/Steps.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply framework/Steps.py:738]: Okay, at least there is a fixme comment.

**Review by joshua-cogliati-inl — COMMENTED** [review]
Changes addressed.

**Review by joshua-cogliati-inl — DISMISSED** [review]

File: framework/JobHandler.py [hunk 9]
```diff
@@ -530,39 +556,60 @@ def getFinished(self, removeFinished=True, jobIdentifier = '', uniqueHandler = "
...
-    ## end with self.__queueLock
-
+      ## check if batches are ready to be returned
+      for groupId in list(self.__batching.keys()):
+        if removeFinished:
+          if len(self.__batching[groupId]['finished']) ==  self.__batching[groupId]['size']:
+            doneBatch = self.__batching.pop(groupId)
+            finished.append(doneBatch['finished'])
+        else:
+          if len(self.__batching[groupId]['finished']) >  self.__batching[groupId]['size']:
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/JobHandler.py]: addressed

File: framework/Models/Model.py [hunk 10]
```diff
@@ -133,7 +133,11 @@ class cls.
   validateDict['Optimizer'][0]['class'       ] ='Optimizers'
   validateDict['Optimizer'][0]['required'    ] = False
   validateDict['Optimizer'][0]['multiplicity'] = 1
-  validateDict['Optimizer'][0]['type']         = ['SPSA','FiniteDifference','ConjugateGradient','SimulatedAnnealing']
+  validateDict['Optimizer'][0]['type']         = ['SPSA',
+                                                  'FiniteDifference',
+                                                  'ConjugateGradient',
+                                                  'SimulatedAnnealing',
+                                                  'GeneticAlgorithm']
```
Comment (joshua-cogliati-inl) [inline framework/Models/Model.py:140]: Adding GeneticAlgorithm probably should wait until GA is merged.

**Review by joshua-cogliati-inl — APPROVED** [review]
I have reviewed the changes and I approve.

File: framework/Models/Model.py (hunk 10 above)
Comment (alfoa) [reply framework/Models/Model.py:140]: addressed

**Comment by alfoa** [comment]
@joshua-cogliati-inl  i addressed last comment

**Comment by mandd** [comment]
i had invalidated again , just to make sure it is a coincidence

**Comment by mandd** [comment]
The passed now

**Pull request merged**